floss/identify.py

```diff
@@ -0,0 +1,130 @@
+# Copyright (C) 2017 FireEye, Inc. All Rights Reserved.
+
+import logging
+import operator
+import collections
+
+import tqdm
+import viv_utils
+
+from floss.features.extract import (
+    abstract_features,
+    extract_insn_features,
+    extract_function_features,
+    extract_basic_block_features,
+)
+from floss.features.features import Arguments, BlockCount, InstructionCount
+
+logger = logger = logging.getLogger(__name__)
+
+
+def get_function_api(f):
+    ret_type, ret_name, call_conv, func_name, args = f.vw.getFunctionApi(int(f))
+
+    return {
+        "ret_type": ret_type,
+        "ret_name": ret_name,
+        "call_conv": call_conv,
+        "func_name": func_name,
+        "arguments": args,
+    }
+
+
+def get_function_meta(f):
+    meta = f.vw.getFunctionMetaDict(int(f))
+
+    return {
+        "api": get_function_api(f),
+        "size": meta.get("Size", 0),
+        "block_count": meta.get("BlockCount", 0),
+        "instruction_count": meta.get("InstructionCount", 0),
+    }
+
+
+def get_max_calls_to(vw, skip_thunks=True, skip_libs=True):
+    calls_to = set()
+
+    for fva in vw.getFunctions():
+        if skip_thunks and is_thunk_function(vw, fva):
+            continue
+
+        # TODO skip_libs and is_library_function
+        #     continue
+
+        calls_to.add(len(vw.getXrefsTo(fva)))
+
+    return max(calls_to)
+
+
+def is_thunk_function(vw, function_address):
+    return vw.getFunctionMetaDict(function_address).get("Thunk", False)
+
+
+def get_function_score_weighted(features):
+    return sum(feature.weighted_score() for feature in features) / sum(feature.weight for feature in features)
+
+
+def find_decoding_functions(vw, functions, disable_progress=None):
+    decoding_candidate_functions = collections.defaultdict(float)
+
+    pbar = tqdm.tqdm
+    if disable_progress:
+        # do not use tqdm to avoid unnecessary side effects when caller intends
+        # to disable progress completely
+        pbar = lambda s, *args, **kwargs: s
+
+    # TODO see FLIRT below
+    # n_funcs = len(functions)
+
+    functions = sorted(functions)
+
+    pb = pbar(functions, desc="finding decoding functions", unit=" functions", postfix="skipped 0 library functions")
+    for f in pb:
+        function_address = int(f)
+
+        if is_thunk_function(vw, function_address):
+            continue
+
+        # TODO ignore library functions via FLIRT
```

Comment:
I'm excited to plug FLIRT into here.

I suspect we should move some of the code around it from capa into viv-utils since they'll work identically.